### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by kibertoad**

### CODE REVIEW

**Comment by kibertoad** [comment]
@lokankr @RonnyLV Please review!

File: README.md [hunk 1]
```diff
@@ -1,2 +1,53 @@
 # undici-retry
 Library for handling retry logic with undici HTTP client
+
+## Basic example
+
+```ts
+import { sendWithRetry } from 'undici-retry';
+import type { RetryConfig } from 'undici-retry'
+import { Client } from "undici";
+import type { Dispatcher } from "undici";
```
Comment (lokankr) [inline README.md:10]: 🟢 
```suggestion
import { Client } from 'undici';
import type { Dispatcher } from 'undici';
```

File: README.md [hunk 2]
```diff
@@ -1,2 +1,53 @@
...
+
+const client = new Client('http://my-url.com', {})
+const request: Dispatcher.RequestOptions = {
+    method: 'GET',
+    path: '/',
+    bodyTimeout: 500,
+    headersTimeout: 500,
+}
+
+const retryConfig = {
```
Comment (lokankr) [inline README.md:20]:
```suggestion
const retryConfig: RetryConfig = {
```

File: lib/undiciRetry.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            statusCode: response.statusCode,
+          },
+        }
+      }
+
+      // retry
+      // undici response body always has to be processed or discarded
+      await response.body.dump()
+    } catch (err: any) {
+      // on internal client error we can't do much; if there are still retries left, we retry, if not, we rethrow an error
```
Comment (lokankr) [inline lib/undiciRetry.ts:75]: `ResponseError` thrown by `resolveBody()` might get swallowed here. But perhaps that's alright?

File: lib/undiciRetry.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            body: resolvedBody,
+            headers: response.headers,
+            statusCode: response.statusCode,
+          },
+        }
+      }
+
+      // Do not retry, return last error response
+      if (
+        retryConfig.statusCodesToRetry.indexOf(response.statusCode) === -1 ||
```
Comment (lokankr) [inline lib/undiciRetry.ts:58]:
```suggestion
        !retryConfig.statusCodesToRetry.includes(response.statusCode) ||
```

File: lib/undiciRetry.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    if (retryConfig.delayBetweenAttemptsInMsecs > 0) {
+      await setTimeout(retryConfig.delayBetweenAttemptsInMsecs)
+    }
+  }
+}
+
+async function resolveBody(response: Dispatcher.ResponseData, safeParseJson = false) {
+  // There can never be multiple content-type headers, see https://www.rfc-editor.org/rfc/rfc7230#section-3.2.2
+  const contentType = response.headers['content-type'] as string | undefined
+  if (contentType?.startsWith('application/json')) {
```
Comment (lokankr) [inline lib/undiciRetry.ts:94]: Nitpick: You could invert this if() and safe a level of indentation.

**Review by lokankr — APPROVED** [review]
Looks great! Only minor comments 🚀

File: lib/undiciRetry.ts (hunk 4 above)
Comment (kibertoad) [reply lib/undiciRetry.ts:58]: It is slightly slower, and for a low-level library it makes sense to microoptimize.
I can generate some benchmarks later today

Comment (lokankr) [reply lib/undiciRetry.ts:58]: Ah gotcha, I just like shiny new things. But good to know 👍

File: lib/undiciRetry.ts (hunk 5 above)
Comment (kibertoad) [reply lib/undiciRetry.ts:94]: This is the hot path, so I prefer to start with it, but yeah, no big difference here

**Pull request merged**